probe-rs/src/debug/debug_step.rs

```diff
@@ -182,145 +174,100 @@ impl SteppingMode {
+                //                 an explicit call to a function, or something the compiler injected, like a `drop()`,
+                //          (a.ii) An interrupt handler diverted the processing.
+                //   (b) We hit a PC at the address of the identified next instruction location,
+                //       which means there was nothing to step into, so the target is now halted (correctly) at the next statement.
+                let instruction_sequence =
+                    InstructionSequence::for_address(debug_info, program_counter)?;
+                let target_pc = match instruction_sequence
+                    .get_first_breakpoint(program_counter.checked_add(1).unwrap_or(program_counter))
                 {
-                    let inclusive_range = current_source_statement.instruction_range.start
-                        ..=current_source_statement.instruction_range.end;
-                    let (core_status, new_pc) = step_to_address(inclusive_range, core)?;
-                    if new_pc == current_source_statement.instruction_range.end {
-                        // We have halted at the address after the current statement, so we can conclude there was no branching calls in this sequence.
-                        tracing::debug!("Stepping into next statement, but no branching calls found. Stepped to next available statement.");
-                    } else if new_pc < current_source_statement.instruction_range.end
-                        && matches!(core_status, CoreStatus::Halted(HaltReason::Breakpoint(_)))
-                    {
-                        // We have halted at a PC that is within the current statement, so there must be another breakpoint.
-                        tracing::debug!(
-                            "Stepping into next statement, but encountered a breakpoint."
-                        );
-                    } else {
-                        // We have reached a location that is not in the current statement range (branch instruction or breakpoint in an interrupt handler).
-                        tracing::debug!(
-                            "Stepping into next statement at address: {:#010x}.",
-                            new_pc
-                        );
+                    Ok(identified_next_breakpoint) => identified_next_breakpoint.address,
+                    Err(DebugError::IncompleteDebugInfo { .. }) => {
+                        // There are no next statements in this sequence, so we will use the return address as the target.
+                        if let Some(return_address) = return_address {
+                            return_address
+                        } else {
+                            return Err(DebugError::IncompleteDebugInfo {
+                                message:
+                                    "Could not determine a step in target. Please use step over."
```

Comment:
This statement could use a few quotes, it took me a minute to parse.

Reply:
Good catch ... thanks :)